### ISSUE TITLE
7$: Error with nice message when trying to use TypeScript-style `const` fields

### ISSUE THREAD

**Issue by strager** [quick-lint/quick-lint-js#235]
```js
class C {
  const f = null;
}
```
quick-lint-js says:
```
hello.js:2:3: error: unexpected token [E054]
```
It should say something like this instead:
```
hello.js:2:3: error: const fields are only allowed in TypeScript, not JavaScript [E???]
```

**Comment by imdarkk** [comment]
I claim this for-hire task. I expect payment after I complete this task. I will email the quick-lint-js team if I am assigned this task.

It might take me some time since I am just getting to know C++

**Comment by imdarkk** [comment]
What file would I have to look into?

**Comment by strager** [comment]
> What file would I have to look into?

`parse_and_visit_class_member` in parse.h

**Comment by strager** [comment]
> I will email the quick-lint-js team if I am assigned this task.

I haven't received your email yet.

**Comment by dlarocque** [comment]
Hey, @strager, I claim this for-hire task. I expect payment after I complete this task. I will email the quick-lint-js team if I am assigned this task

I'd like to give my first issue a try!

**Comment by dlarocque** [comment]
@strager I've been struggling on this quite a bit.

So far, I've got this much:
- A new error type for TypeScript style const fields: https://github.com/dlarocque/quick-lint-js/commit/1ffae0f64f3994de2c2c8d8a1372e58927e82916
- A really simple new case in `parse_and_visit_class_member` that reports an error when a const keyword is parsed in a class https://github.com/dlarocque/quick-lint-js/commit/267231e9c288c92aa477b1cda8a795dbb343f4b0

For the example you gave when you opened this issue, I get this error when running the linter: 
`main.js:2:11: error: const fields within classes are only allowed in TypeScript, not Javascript [E165]`

Two tests fail in `test_parse-class.cpp` , those are 
`class_fields_without_initializer_allow_asi_after_name`, and
`class_statement_with_keyword_property`.  

After looking at these tests, my guess is that the order of visits(?) is incorrect because I'm messing up the order of parsing (I don't really know what I'm saying here)  I have a few questions about this, but I think it might be better if I ask them on stream tonight.

Sorry about the slow progress on this

**Comment by strager** [comment]
Discussed in person.

Interesting piece of code: https://github.com/dlarocque/quick-lint-js/blob/267231e9c288c92aa477b1cda8a795dbb343f4b0/src/quick-lint-js/parse.h#L1489

Test cases:
```js
// needs test
class C {
  const f = null; // invalid JS; valid TS
}
// needs test
class C2 {
  const     // valid JS
  f = null; // valid JS
}

// already tested
class E {
  const  // valid JS
}
class E2 {
  const() { } // valid JS
}
```

**Comment by dlarocque** [comment]
@strager 

- I added the error message, all tests pass, but the way I did it sucks 3f674d171208fae0aa68059073f8a837b9cfa0f6
- I added documentation for the new error type 841b8261e8cdae4be18c7e7c61f1def8540c27e4

I'm wondering if there are any other test cases I should write tests for other than the ones here?

> Test cases:
> 
> ```js
> // needs test
> class C {
>   const f = null; // invalid JS; valid TS
> }
> // needs test
> class C2 {
>   const     // valid JS
>   f = null; // valid JS
> }
